packages/teleport/src/components/FilterByLabelList/FilterByLabelList.tsx

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2021 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useState, useMemo } from 'react';
+import { useLocation } from 'react-router';
+import { Option } from 'shared/components/Select';
+import SelectFilter from 'teleport/components/SelectFilter';
+import history from 'teleport/services/history';
+
+export default function FilterableList({
+  data = [],
+  TableComponent,
+  ...tableProps
+}: Props) {
+  const { search, pathname } = useLocation();
+  const labelsFromQuery = useMemo<Option[]>(() => {
+    const searchParams = new URLSearchParams(search);
+    const query = searchParams.get('labels');
+    if (!query || query === 'null') {
+      return [];
+    }
+
+    return decodeURIComponent(query)
+      .split(',')
+      .map(label => ({ value: label, label }));
+  }, []);
+
+  const labels = useMemo<Option[]>(() => makeLabelOptions(data), []);
+  const [selectedLabels, setSelectedLabels] = useState<Option[]>(
+    labelsFromQuery
+  );
+  const [filteredData, setFilteredData] = useState(() =>
+    filterData(data, labelsFromQuery)
+  );
+
+  function onFilterApply(labels: Option[]) {
+    const filtered = filterData(data, labels);
+
+    setFilteredData(filtered);
+    setSelectedLabels(labels);
+    updateUrlQuery(labels, pathname);
+  }
```

Comment:
I'm thinking about this and we can simplify things a bit:
- `selectedLabels` are duplicating the state from `labelsFromQuery` so it can be removed 
- `filteredData` can be `useMemo` calculating new result whenever `data` or `labelsFromQueryChange`
- `onFilterApply` can be much simpler and only call `updateUrlQuery()`
Thanks to these changes URL will be our single source of truth and it will be always in sync with the UI.

Let me know what do you think about this :)

Reply:
i think it's a great idea :) i thought about this too, but i hesitated to change b/c we are basically reformatting labels to Option type, when it was originally a Option type (it's also harder to play around with it in storybook). Though it's probably a pre-optimization thinking and doesn't matter too much

Reply:
I'm not sure if I understand, you mean changing `tags` from list of objects (`{ name: 'env', value: 'prod' }`) to list of strings (`env: prod`) in storybook? But both now and before we had list of strings in `FilterByLabelList`, right?

Reply:
I'm not sure if I understand, you mean changing tags from list of objects ({ name: 'env', value: 'prod' }) to list of strings (env: prod) in storybook? But both now and before we had list of strings in FilterByLabelList, right?

Reply:
Prior, I was manually updating labels: `setSelectedLabels(newLabels)`. Now we rely on the url change, but I am not sure how to configure storybook so that this url change can be picked up. 

instead of (what we currently have):
```ts
const selectedLabels = useMemo<Option[]>(() => {
    const searchParams = new URLSearchParams(search);
    const query = searchParams.get('labels');
    if (!query || query === 'null') {
      return [];
    }

    return decodeURIComponent(query)
      .split(',')
      .map(label => ({ value: label, label }));
}, [search]);

function onFilterApply(labels: Option[]) {
    updateUrlQuery(labels, pathname);
}
```

equivalent, but also works well with storybook:
```ts
const [selectedLabels, setSelectedLabels] = useState<Option[]>(() => {
    const searchParams = new URLSearchParams(search);
    const query = searchParams.get('labels');
    if (!query || query === 'null') {
      return [];
    }

    return decodeURIComponent(query)
      .split(',')
      .map(label => ({ value: label, label }));
});

function onFilterApply(labels: Option[]) {
    setSelectedLabels(labels)
    updateUrlQuery(labels, pathname);
}
```

Reply:
Yeah, it doesn't work in storybook because `teleport history service` doesn't operate on history of `MemoryRouter`. Maybe we can use `const history = useHistory()`? It picks up the current history object.